torch/nested/_internal/ops.py

```diff
@@ -847,28 +843,82 @@ def is_same_size_default(func, *args, **kwargs):
 
 
 @register_jagged_func(
-    torch.ops.aten.sum.dim_IntList, "self: jt, dim: any?, keepdim: any?, dtype: any?"
+    torch.ops.aten.sum.dim_IntList,
+    "self: jt_all, dim: any?, keepdim: any?, dtype: any?",
 )
 def sum_dim_IntList(func, *args, **kwargs):
-    # sum_dim_IntList can produce a NT or a T depending on whether the ragged dims
-    # are reduced away.
+    """
+    Performs a sum along the provided tensor dimension.
+    Returns a dense tensor if the ragged dimension is reduced away, else returns a nested tensor.
+    """
     _, new_kwargs = normalize_function(
         func, args=args, kwargs=kwargs, normalize_to_only_use_kwargs=True
     )
     inp = new_kwargs.pop("input")
-    assert inp._ragged_idx == 1
-    new_kwargs["dim"], ragged_reduced_away = _wrap_jagged_dims(
-        inp.dim(), new_kwargs["dim"], "sum"
+
+    (
+        new_kwargs["dim"],
+        reduce_on_batch,
+        reduce_on_ragged,
+        reduce_on_non_batch,  # noqa: UFMT
+    ) = _wrap_jagged_dims(
+        inp.dim(),
+        new_kwargs["dim"],
+        "sum",
+        inp._ragged_idx,
     )
 
-    if not ragged_reduced_away:
-        return NestedTensor(func(inp._values, **new_kwargs), **extract_kwargs(inp))
-    else:
-        # Don't wrap because we reduced away the raggedness
-        out = func(inp._values, **new_kwargs)
+    if inp._ragged_idx != 1:
+        raise RuntimeError("sum(): not supported when ragged_idx != 1 for NestedTensor")
+
+    if reduce_on_ragged and inp._lengths is not None:
+        raise RuntimeError(
+            "sum(): not supported for nested tensors with holes, i.e. when lengths is not None, "
+            + "if reducing across the ragged dimension for NestedTensor"
+        )
+
+    if reduce_on_ragged:  # raggedness reduced away --> return dense tensor
```

Comment:
minor nit: the branching here is a bit confusing IMO, but I also don't really have a suggestion on how to make it cleaner haha

Reply:
agreed! i think it was worse before and now it's a bit more cleaned up, but it's still tedious to read unfortunately. one possible solution is to create individual boolean conditions for each reduction case, but this could also make the code long-winded to read